### ISSUE TITLE
feat(barcode-scanning): zoom functionality

### ISSUE THREAD

**Issue by Melynt3** [capawesome-team/capacitor-mlkit#72]
### Plugin(s)

- [X] Barcode Scanning
- [ ] Face Detection
- [ ] Translation

### Current problem

Hi,

I asked about this a few days ago in a different topic, but I would like to officially request a new feature, which is zoom functionality. It's a big challenge to work with small barcodes. I encountered a problem when scanning small CODE128 codes; basically, the scanner does not read them. I was able to hardcode a workaround in the library itself by adding a zoom of 1.6f during camera initialization, but this is only a temporary solution. It would be a great feature that would make the library even more attractive.

### Preferred solution

Adding a zoom based on the google barcode scanner, which is implemented in the demo version
either zoom by "enlarging the screen" or by adding a slider

### Alternative options

While working on react library i occured this kind of zoom which was very intuitive and it was known to be there. 
![1 (3)](https://github.com/capawesome-team/capacitor-mlkit/assets/45972864/53e28e97-e47c-44ad-82be-70f638717e93)

**Comment by sephallen** [comment]
Having the ability to set the zoom level on initialisation would be very useful.

I am also experiencing similar issues with small CODE39 barcodes, the library is not able to read the value unless you pinch to zoom around 200%.

**Comment by Melynt3** [comment]
Yup, i occured the same error with different type. 

i temporary fixed it by setting zoom lvl to 1.65 ratio in library from android studio, if u have more efficient way to manage it i would be greatful. 


```
CameraControl cameraControl = camera.getCameraControl();
float zoomLevel = 1.65f;
cameraControl.setZoomRatio(zoomLevel);
```

**Comment by robingenz** [comment]
PRs are welcome. I propose these types:

```ts
  /**
   * Set the zoom ratio of the camera.
   * 
   * Only available on Android and iOS.
   */
  setZoomRatio(): Promise<void>;
  /**
   * Get the zoom ratio of the camera.
   * 
   * Only available on Android and iOS.
   */
  getZoomRatio(): Promise<void>;
  /**
   * Get the minimum zoom ratio of the camera.
   * 
   * Only available on Android and iOS.
   */
  getMinZoomRatio(): Promise<void>;
  /**
   * Get the maximum zoom ratio of the camera.
   * 
   * Only available on Android and iOS.
   */
  getMaxZoomRatio(): Promise<void>;
```

Docs:
- https://developer.android.com/reference/androidx/camera/core/CameraControl#setZoomRatio(float)
- https://developer.android.com/reference/androidx/camera/core/ZoomState#getZoomRatio()
- https://developer.apple.com/documentation/avfoundation/avcapturedevice/1624611-videozoomfactor
- https://developer.apple.com/documentation/avfoundation/avcapturedevice/2865758-maxavailablevideozoomfactor
- https://developer.apple.com/documentation/avfoundation/avcapturedevice/1624614-ramp

**Comment by robingenz** [comment]
I am setting a [bounty](https://capawesome.io/contributing/bounties/) of **$50** on this issue.

https://github.com/capawesome-team/capacitor-mlkit/issues/72#issuecomment-1759279257 describes what needs to be done.
The changes must be fully tested.
I recommend to make the changes in the `definitions.ts` file first and let me review them before starting with the actual implementation.

Let me know if you have any questions.

**Comment by vosecek** [comment]
Hi, any chance this task could be implemented?

**Comment by robingenz** [comment]
@vosecek Sure, but there is no ETA yet.
[Here](https://capawesome.io/contributing/feature-requests/#after-creating-an-issue) you can find information on how to speed up the process.

**Comment by vosecek** [comment]
@robingenz what is appropriate price/bounty for implementation?

**Comment by robingenz** [comment]
@vosecek You can set any amount. Just take care of the minimum and maximum amount (see [FAQ](https://capawesome.io/contributing/bounties/create-a-bounty/#is-there-a-minimum-bounty-amount)).

**Comment by vosecek** [comment]
@robingenz Contribution 200$ sent https://opencollective.com/capawesome/contributions/720755

**Comment by robingenz** [comment]
@vosecek Thank you for your contribution! I will prioritize the issue.

**Comment by robingenz** [comment]
I’ve just published a dev version, see https://github.com/capawesome-team/capacitor-mlkit/pull/112#issuecomment-1865041507.
Feel free to give it a try. I would appreciate any feedback.

**Comment by vosecek** [comment]
> I’ve just published a dev version, see [#112 (comment)](https://github.com/capawesome-team/capacitor-mlkit/pull/112#issuecomment-1865041507). Feel free to give it a try. I would appreciate any feedback.

Works great, thanks!

`ion-range` is not handy component for this, cause zoom is needed only 2x-4x, using sliding range is difficult to set useful zoom ratio.

My implementation for controls is 
```
<ion-fab slot="fixed" horizontal="end" vertical="center">
        <ion-fab-button (click)="increaseZoom()" [disabled]="currentZoomRatio >= maxZoomRatio" class="ion-margin-bottom" [color]="'primary'">
            <ion-icon size="large" name="search-outline"></ion-icon>
        </ion-fab-button>
        <ion-fab-button (click)="decreaseZoom()" [disabled]="currentZoomRatio <= minZoomRatio" [color]="'dark'">
            <ion-icon size="small" name="search-outline"></ion-icon>
        </ion-fab-button>
    </ion-fab>
```

```
// set this.currentZoomRatio on init
increaseZoom() {
    // @ts-ignore
    this.setZoomRatio({detail: {value: this.currentZoomRatio + 1}});
  }

  decreaseZoom() {
    // @ts-ignore
    this.setZoomRatio({detail: {value: this.currentZoomRatio - 1}});
  }
```
This is just my suggestion for demo code. Anyway, this is great and fulfil our needs.

**Comment by Melynt3** [comment]
I managed to implement it simply on buttons but with two ways of working, both methods on one buttons.

1. Zoom on hold, which simply zoom while we holding "Zoom in" button
2. Zoom on click, as it says :D Zooming by 20% on click 

Little be overcooked but works fine for me. 
ill try to implement it on some good looking **ion-range** in the future.

`ts
	zoomRatio: number = 1.0;
	zoomInPressing: boolean = false;
	cleaner: Subject<void> = new Subject<void>();
	zoomIntervalSubscription: Subscription | undefined;
	
	
		ngOnDestroy(): void {
		this.cleaner.next();
		this.cleaner.complete();
	}

	startZoomIn(): void {
		this.zoomInPressing = true;

		this.zoomIntervalSubscription = interval(25)
			.pipe(takeUntil(this.cleaner))
			.subscribe(async () => {
				if (this.zoomInPressing) {
					this.zoomRatio += 0.01;
					await this.setZoomRatio();
				}
			});
	}

	stopZoomIn(): void {
		this.zoomInPressing = false;
		if (this.zoomIntervalSubscription) {
			this.zoomIntervalSubscription.unsubscribe();
		}
	}

	async zoomIn(): Promise<void> {
		this.zoomRatio += 0.2;
		await this.setZoomRatio();
		this.stopZoomIn();
	}

	async zoomOut(): Promise<void> {
		const newZoomRatio = this.zoomRatio - 0.2;
		this.zoomRatio = Math.max(newZoomRatio, 1);
		await this.setZoomRatio();
		this.stopZoom();
	}

	startZoomOut(): void {
		this.zoomIntervalSubscription = interval(25)
			.pipe(takeUntil(this.cleaner))
			.subscribe(async () => {
				const newZoomRatio = this.zoomRatio - 0.01;
				this.zoomRatio = Math.max(newZoomRatio, 1);
				await this.setZoomRatio();
			});
	}

	stopZoom(): void {
		if (this.zoomIntervalSubscription) {
			this.zoomIntervalSubscription.unsubscribe();
		}
	}

	async setZoomRatio(): Promise<void> {
		const options: SetZoomRatioOptions = {
			zoomRatio: this.zoomRatio
		};
		await BarcodeScanner.setZoomRatio(options);
	}
`


`html
<ion-row class="zoom-controls">
                <ion-col size="4" class="ion-text-center">
                    <ion-button
							(click)="zoomIn()"
                            (touchstart)="startZoomIn()"
                            (touchend)="stopZoomIn()"
                            fill="clear" color="light">
                        <ion-icon name="add" slot="end"></ion-icon>
                        Zoom in
                    </ion-button>
                </ion-col>

                <ion-col size="4" class="ion-text-center">
                    <ion-text color="light">
                        {{ zoomRatio >= 1 ? Math.floor(zoomRatio * 100) + '%' : '100%' }}
                    </ion-text>
                </ion-col>

                <ion-col size="4" class="ion-text-center">
                    <ion-button
							(click)="zoomOut()"
                            (touchstart)="startZoomOut()"
                            (touchend)="stopZoom()"
                            fill="clear" color="light">
                        <ion-icon name="remove" slot="end"></ion-icon>
                        Zoom out
                    </ion-button>
                </ion-col>
            </ion-row>
`
